field_mask/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+            children_fields(&mask, "a"),
+            vec!["b.c".to_string(), "b.d".to_string()]
+        );
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[derive(thiserror::Error, Debug)]
+    enum Error {
+        #[error("bad something")]
+        BadSomething,
+        #[error("field validation")]
+        FieldViolation(#[from] FieldViolation),
+    }
+
+    type Result<T> = std::result::Result<T, Error>;
+
+    #[derive(Clone, Debug)]
+    struct Config {
+        id: u32,
+        something: Option<Foo>,
+        other: Option<Foo>,
+    }
+
+    #[derive(Clone, Debug)]
+    struct Foo {
+        bar: u32,
+        baz: u32,
+    }
+
+    fn process_id(id: u32) -> Result<u32> {
+        Ok(1 + id)
+    }
+
+    fn process_something(something: &Foo, mask: FieldMask) -> Result<String> {
+        Ok(format!("{} inner mask {:?}", something.bar, mask.paths))
+    }
+
+    fn process_something_err(_something: &Foo, _mask: &FieldMask) -> Result<()> {
+        Err(Error::BadSomething)
+    }
+
+    fn test_config() -> Config {
+        Config {
+            id: 42,
+            something: Some(Foo { bar: 1, baz: 2 }),
+            other: None,
+        }
+    }
+
+    #[test]
+    fn test_match_field_mask_ok() {
+        fn test() -> Result<()> {
+            let cfg = test_config();
+            let mask = FieldMask {
+                paths: vec!["something.other".to_string(), "id".to_string()],
+            };
+
+            match_field_mask!((cfg, mask) {
+                .id => process_id(*id)?,
+                .something if let Some(something) |(mask) => process_something(&something, mask)?,
+                .other if let Some(other) => println!("{}", 1 + other.bar),
+            });
+            Ok(())
+        }
+        test().unwrap();
+    }
+
+    #[test]
+    // This test just ensures that the macro can cope with parameters being
+    // references.
+    fn test_match_field_ref() {
+        fn test() -> Result<()> {
+            let cfg = test_config();
+            let mask = FieldMask {
+                paths: vec!["id".to_string()],
+            };
+
+            match_field_mask!((&cfg, &mask) {
+                .id => process_id(*id)?,
+            });
+            Ok(())
+        }
+        test().unwrap();
+    }
+
+    #[test]
+    fn test_match_field_mask_required() {
+        fn test() -> Result<()> {
+            let cfg = test_config();
+            let mask = Some(FieldMask { paths: vec![] });
+
+            match_field_mask!((cfg,mask) {
+                .id => process_id(*id)?,
+            });
+
+            Ok(())
+        }
+        test().unwrap();
+    }
+
+    #[test]
+    fn test_match_field_mask_required_err() {
+        fn test() -> Result<()> {
+            let cfg = test_config();
+            let mask = None;
+
+            match_field_mask!((cfg, mask) {
+                .id => process_id(*id)?,
+            });
+
+            Ok(())
+        }
+        test().expect_err("error");
```

Comment:
I think this test and the others that `expect_err` would be clearer and more valuable if they were something like:

```suggestion
        assert!(matches!(test(), Err(FieldViolation::required("update_mask"))));
```

or whatever actually compiles, hehe :)

Reply:
my personal favorite is `assert_eq!(test().unwrap_err().to_string(), "The actual error you expect here so you make sure the message makes sense")`

Reply:
thanks for the suggestion! I opted for matching the actual error type instead of the `to_string()` but mostly because the FieldViolation to_string implementation doesn't seem to render the full description (and I reached my yak shaving recursion limit for this week).